### PR TITLE
Include node-external-ip in serving-kubelet.crt SANs

### PR DESCRIPTION
**Pull request by alexey-medvedchikov**
#### Problem ####

When kube-apiserver is set up to reach kubelet API via ExternalIP it fails because the address is not included into certificate SANs.

#### Proposed Changes ####

Include node-external-ip in serving-kubelet.crt SANs to make sure kubelet API (serving-kubelet.crt) TLS certificate passes validation on kube-apiserver side.

#### Types of Changes ####

New Feature

#### Verification ####

* Deploy machine with supplied `--node-external-ip` argument set to machine external IP address.
* Check generated kubelet certificate on the machine: `openssl x509 -in /var/lib/rancher/k3s/agent/serving-kubelet.crt -text -noout | grep -A 3 'Subject Alternative Name'`. It must include external IP address in the list of SANs.

To verify that the problem is solved you need to create 2 machines with external addresses absent on the local network interfaces, i.e. externalIP provided via NAT, e.g. GCP or AWS EC2. The machines must be located in separate networks, so they won't be able to talk to each other through internal network.

* Server: `k3s server --token $TOKEN --kube-apiserver-arg kubelet-preferred-address-types=ExternalIP --node-external-ip $EXTERNAL_IP_1`
* Node: `k3s agent --token $TOKEN --server https://$EXTERNAL_IP1:6443 --node-external-ip $EXTERNAL_IP_2`
* Get kubeconfig

After successful installation `kubectl logs ...` and `kubectl exec ...` should work.

#### Linked Issues ####

#4520

#### Further Comments ####

I also did some refactoring to DRY the code used to parse IP addresses, this caused error log lines to change slightly:
* In parsing `node-external-ip` error: `invalid node-external-ip %s` -> `invalid node-external-ip: invalid ip format '%s'`
* In parsing `node-ip` error: `invalid node-ip %s` -> `invalid node-ip: invalid ip format '%s'`

Also, added unit tests for the IP parsing code.

### CODE REVIEW

File: pkg/util/net_test.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
+package util
+
+import (
+	"net"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
```
Comment (dereknola) [inline pkg/util/net_test.go:7]: Thanks for the unit test, but please follow the guidelines for unit tests https://github.com/k3s-io/k3s/blob/master/tests/TESTING.md#framework. You don't need to rewrite much, just the assertions. We don't use the testify assert package on purpose, we use regular comparisons or DeepEqual with an error. See https://github.com/k3s-io/k3s/blob/master/pkg/configfilearg/parser_test.go#L335 for an example.

**Review by dereknola — CHANGES_REQUESTED** [review]

File: pkg/util/net_test.go (hunk 1 above)
Comment (alexey-medvedchikov) [reply pkg/util/net_test.go]: @dereknola Sorry about that, didn't pay enough attention. Changed the code to be conformant with the guidelines.

**Review by dereknola — APPROVED** [review]

File: pkg/agent/config/config.go [hunk 2]
```diff
@@ -352,6 +352,11 @@ func get(ctx context.Context, envInfo *cmds.Agent, proxy proxy.Proxy) (*config.N
 		return nil, err
 	}
 
+	nodeExternalIPs, err := util.ParseStringSliceToIPs(envInfo.NodeExternalIP)
+	if err != nil {
+		return nil, fmt.Errorf("invalid node-external-ip: %w", err)
+	}
+
```
Comment (brandond) [inline pkg/agent/config/config.go:359]:
```suggestion
	nodeIPs = append(nodeIPs, nodeExternalIPs...)

```

File: pkg/agent/config/config.go [hunk 3]
```diff
@@ -362,7 +367,8 @@ func get(ctx context.Context, envInfo *cmds.Agent, proxy proxy.Proxy) (*config.N
 
 	os.Setenv("NODE_NAME", nodeName)
 
-	servingCert, err := getServingCert(nodeName, nodeIPs, servingKubeletCert, servingKubeletKey, newNodePasswordFile, info)
+	nodeExternalAndInternalIPs := append(nodeIPs, nodeExternalIPs...)
+	servingCert, err := getServingCert(nodeName, nodeExternalAndInternalIPs, servingKubeletCert, servingKubeletKey, newNodePasswordFile, info)
```
Comment (brandond) [inline pkg/agent/config/config.go:371]: Don't change this here, just add the external IPs to the nodeIPs list up above.

**Review by brandond — CHANGES_REQUESTED** [review]

**Review by brandond — APPROVED** [review]

**Pull request merged**